### PR TITLE
Add features to comments section functionality.

### PR DESCRIPTION
**Pull request by keelywan**
## Before
comments

## After 
comments4

## Update
- Reformat select tags 
- Add margins between inputs 
comments5

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -19,30 +19,56 @@
...
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("date", SortDirection.DESCENDING);
+    String num = request.getParameter("num");
+    String order = request.getParameter("order");
+    String user = request.getParameter("user").trim().toUpperCase();
+    
+    int displayNum = Integer.MAX_VALUE;
+    if(!num.equals("all")) {
+      displayNum = Integer.parseInt(num);
```
Comment (cftorres) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: For one line conditionals like this and below, you could try using the ternary operator.  Condenses the code a bit.  So it would look something like 
`int displayNum = num.equals("all") ? Integer.MAX_VALUE : Integer.parseInt(num)`

In case you're not familiar, you can read about it here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -19,30 +19,56 @@
...
+
+    Query query = new Query("Comment").addSort("date", sort);
+
+    if(user != null && !user.equals("")) {
+      query = new Query("Comment").addSort("date", sort).setFilter(new FilterPredicate("username", FilterOperator.EQUAL, user));
+    }
 
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
+    int total = results.countEntities(FetchOptions.Builder.withLimit(Integer.MAX_VALUE));
```
Comment (cftorres) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Nit: I might out this line right before you use it below because it's easier to follow when the variable is consumed right after declaration

File: portfolio/src/main/java/com/google/sps/data/CommentRequest.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+/** Class containing requested comment data. */ 
+public class CommentRequest {
+
+  /** List of requested comments. */
+  private List<Comment> comments;
+
+  /** Total number of comments. */
+  private int total;
+
+  public CommentRequest(List<Comment> com, int num) {
```
Comment (nickKelman) [inline portfolio/src/main/java/com/google/sps/data/CommentRequest.java]: It's good to use descriptive variable names when possible.

It took me a bit to learn what 'num' and 'com' were. The screenshots helped!

perhaps 'commentsToReturn' and 'maxNumberOfComments' to return?

After reading through your code, it seems the user can request a limited number of comments and specific comments and the limited number of comments will set a cut off point for the specific comments. I could also see the possibility of an API like this returning all comments asked for, ignoring the max number. It might be worth specifying this behavior in comments.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -19,30 +19,56 @@
...
+import com.google.appengine.api.datastore.Query.FilterPredicate;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("date", SortDirection.DESCENDING);
+    String num = request.getParameter("num");
```
Comment (nickKelman) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I would try to use a more descriptive variable name here.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -19,30 +19,56 @@
...
+    String order = request.getParameter("order");
+    String user = request.getParameter("user").trim().toUpperCase();
+    
+    int displayNum = Integer.MAX_VALUE;
+    if(!num.equals("all")) {
```
Comment (nickKelman) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I assume all is a keyword you pass along from the client.

I would declare an enum like constant for this value similar to what Andres did here:
https://github.com/andres-holguin/step/commit/289904a1bd14dc78532409f1bd4e1502beb680ad

Comment (nickKelman) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This applies to the sorting option as well.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -19,30 +19,56 @@
...
+
+    SortDirection sort = SortDirection.DESCENDING;
+    if(order.equals("oldest")) {
+      sort = SortDirection.ASCENDING;
+    }
+
+    Query query = new Query("Comment").addSort("date", sort);
+
+    if(user != null && !user.equals("")) {
+      query = new Query("Comment").addSort("date", sort).setFilter(new FilterPredicate("username", FilterOperator.EQUAL, user));
```
Comment (nickKelman) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: this line is long to the point where I need to scroll to read it. Can you try to format your files to set max line length to what I assume is 100 characters.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -19,30 +19,56 @@
...
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (nickKelman) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:46]: This function is doing a lot. Breaking out some of this logic into private functions would aid in readability.

Perhaps some functions like setSortStyle, setFilter, setComments.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
+// Copyright 2019 Google LLC
```
Comment (nickKelman) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java:1]: This commit has a lot going on. This deletion logic seems like it could be broken out into a separate commit and is independent from the filtering and sorting logic.

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -33,20 +33,29 @@ function displayLocation(name) {
...
+
+  const response = await fetch('/data' + num + order + user);
   const content = await response.text();
   var obj = JSON.parse(content); 
 
+  clearDiv();
+
   const commentEl = document.getElementById('comment-container');
-  obj.forEach((line) => {
+  const descriptionSpan = document.createElement('p');
```
Comment (nickKelman) [inline portfolio/src/main/webapp/script.js]: This seems like a paragraph, not a span

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -56,22 +65,92 @@ async function getServerContent() {
...
+
+  const deleteButton = document.createElement('button');
+  deleteButton.innerHTML = '<i class="fas fa-trash-alt"></i>';
+  deleteButton.setAttribute('class', 'del-attr');
+  deleteButton.addEventListener('click', () => {
+    deleteComment(comment);
+  });
 
-  var textP = document.createElement("p"); 
+  const textP = document.createElement('p');
```
Comment (nickKelman) [inline portfolio/src/main/webapp/script.js]: what is 'P' meant to communicate? Parameter?

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -56,22 +65,92 @@ async function getServerContent() {
...
+  divElement.appendChild(deleteButton);
   divElement.appendChild(dateP); 
   divElement.appendChild(textP); 
   return divElement;
 }
+
+/**
+ * Clears div element containing comments.
+ */
+function clearDiv() {
```
Comment (nickKelman) [inline portfolio/src/main/webapp/script.js]: 'removeAllCommentsFromPage' ?

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -56,22 +65,92 @@ async function getServerContent() {
...
+  const commentEl = document.getElementById('comment-container');
+  while(commentEl.firstChild) {
+    commentEl.removeChild(commentEl.firstChild);
+  }
+}
+
+/**
+ * Deletes all comments from Datastore.
+ */
+async function deleteAll() {
```
Comment (nickKelman) [inline portfolio/src/main/webapp/script.js]: 'deleteAllComments' ?

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -56,22 +65,92 @@ async function getServerContent() {
...
+  getServerContent();
+}
+
+/**
+ * Changes text of selected number option then retrieves server content.
+ */
+function changeDisplayNum() {
+  const selectEl = document.getElementById('display-num');
+  var index = selectEl.selectedIndex;
+  for(var i = 0; i < selectEl.options.length; i++) {
```
Comment (nickKelman) [inline portfolio/src/main/webapp/script.js]: s/var/let

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -56,22 +65,92 @@ async function getServerContent() {
...
+    if(index === i) {
+      selectEl[i].innerHTML = 'Show: ' + selectEl[i].innerHTML;
```
Comment (nickKelman) [inline portfolio/src/main/webapp/script.js]: This is a lot of logic to control including the word 'Show' for the default option in the list. Perhaps you can communicate that this option controls what gets shown in the list outside the dropdown option? You could then delete this logic and the function below it.

**Review by nickKelman — COMMENTED** [review]
A lot of great functionality here. Great work.

If possible try to break out independent features into separate changes.

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (keelywan) [reply portfolio/src/main/webapp/script.js]: The 'P' was meant to show that it was a paragraph element, but I can see that's unclear, so I'll change it to textParagraph.

File: portfolio/src/main/webapp/script.js (hunk 14 above)
Comment (keelywan) [reply portfolio/src/main/webapp/script.js]: Deleted!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String requestID = request.getParameter("id");
+
+    if(requestID != null) {
+      // Delete specific comment 
+      long id = Long.parseLong(requestID);
+      Key commentEntityKey = KeyFactory.createKey("Comment", id);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (cftorres) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteDataServlet.java]: This can be turned into an instance variable so you can use the same one below too

File: portfolio/src/main/webapp/script.js [hunk 16]
```diff
@@ -33,20 +33,29 @@ function displayLocation(name) {
...
+  label.style.display = 'none';
   img.style.opacity = 1; 
 }
 
 /**
  * Async function to fetch server content and add it to DOM.
  */
 async function getServerContent() {
-  const response = await fetch('/data');
+  var num = "?num=" + document.getElementById('display-num').value;
```
Comment (cftorres) [inline portfolio/src/main/webapp/script.js]: perhaps this portion could be put in a helper getServerContent() to simplify this function and instead this function could probably more accurately be named something like displayServerContent().  As Nick pointed out above, it could be useful to think about ways to split up meaty functions like this into helper functions and condense your logic.

File: portfolio/src/main/webapp/script.js [hunk 17]
```diff
@@ -56,22 +65,92 @@ async function getServerContent() {
...
+    commentEl.removeChild(commentEl.firstChild);
+  }
+}
+
+/**
+ * Deletes all comments from Datastore.
+ */
+async function deleteAll() {
+  const response = await fetch('/delete-data', {method: 'POST'});
+  getServerContent();
```
Comment (cftorres) [inline portfolio/src/main/webapp/script.js]: Going back to my comment above, I was confused initially reading what this function did because I couldn't understand why you were getting server content after this but I realize it's because you're trying to re-render the (lack of) comments after deletion, so this function could probably use renaming to make clearer its exact behavior.

File: portfolio/src/main/webapp/script.js [hunk 18]
```diff
@@ -56,22 +65,92 @@ async function getServerContent() {
...
+  while(commentEl.firstChild) {
+    commentEl.removeChild(commentEl.firstChild);
+  }
+}
+
+/**
+ * Deletes all comments from Datastore.
+ */
+async function deleteAll() {
+  const response = await fetch('/delete-data', {method: 'POST'});
```
Comment (cftorres) [inline portfolio/src/main/webapp/script.js]: Here and below, do you need to save this in a variable if you're not using it?

File: portfolio/src/main/webapp/script.js (hunk 16 above)
Comment (keelywan) [reply portfolio/src/main/webapp/script.js]: Got it - will split getServerContent() into two functions!

File: portfolio/src/main/java/com/google/sps/data/CommentRequest.java (hunk 3 above)
Comment (keelywan) [reply portfolio/src/main/java/com/google/sps/data/CommentRequest.java]: I'll add a better description in CommentRequest about the different behaviors.

File: portfolio/src/main/webapp/script.js (hunk 18 above)
Comment (keelywan) [reply portfolio/src/main/webapp/script.js]: Nope, I just removed it. I thought at first that I would need to save it to a variable in order for the await fetch to work.

File: portfolio/src/main/java/com/google/sps/data/CommentRequest.java [hunk 19]
```diff
@@ -24,11 +24,17 @@
...
   private List<Comment> comments;
 
-  /** Total number of comments. */
-  private int total;
+  /**
+   * Total number of comments found from the query. If no limit is set and all comments asked for
+   * are returned, total should be equal to the size of List<Comment> comments. Alternatively, if
+   * the limit is greater than the number of comments returned, then total should also be equal to
+   * the size of List<Comment> comments. Otherwise, only a subset of all the comments found will
+   * be returned to the client.
```
Comment (cftorres) [inline portfolio/src/main/java/com/google/sps/data/CommentRequest.java]: From my understanding of your code, I think this is a little inaccurate/misleading.  Correct me if I'm wrong, but based on what I see below of how you create a CommentRequest, `totalComments` always contains the total number of comments, no?  (With the exception of if there are more comments than MAX_INTEGER, then it contains MAX_INTEGER.). So I don't see the need to include all of this conditional explaining in the comment.  I think instead you could just say something like "Total number of comments existing in Datastore" (to make clear that "comments" doesn't refer to your variable above.  Instead, I would include what you're alluding to here in the comment above for the "comments" variable.  

And as a nit, I would say this comment is perhaps a little verbose.  If you're using intuitive variable names, which you now are :) you can avoid having to explain so much in the comment because the reader will be able to intuit more from the code.  So you could probably condense this by saying something like "A list that contains a strict subset of the comments in Datastore if fewer than the total were requested, else all of the comments, with a maximum of MAX_INTEGER."

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 20]
```diff
@@ -40,33 +40,25 @@
...
-    }
-
-    Query query = new Query("Comment").addSort("date", sort);
-
-    if(user != null && !user.equals("")) {
-      query = new Query("Comment").addSort("date", sort).setFilter(new FilterPredicate("username", FilterOperator.EQUAL, user));
-    }
+    int numberOfCommentsDisplayed = setCommentLimit(maxNumberOfComments);
+    SortDirection sortOrder = setSortStyle(orderOfComments);
+    Query query = setQuery(sortOrder, user);
```
Comment (cftorres) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:54]: Great job extracting these out!

File: portfolio/src/main/webapp/script.js [hunk 21]
```diff
@@ -32,48 +32,56 @@ function displayLocation(name) {
...
-async function getServerContent() {
-  var num = "?num=" + document.getElementById('display-num').value;
-  var order = "&order=" + document.getElementById('display-order').value;
-  var user = "&user=" + document.getElementById('display-user').value;
-
-  const response = await fetch('/data' + num + order + user);
-  const content = await response.text();
-  var obj = JSON.parse(content); 
+async function displayServerContent() {
+  const serverResults = await getServerContent();
```
Comment (cftorres) [inline portfolio/src/main/webapp/script.js]: I think you might be able to do const `const {comments, totalComments} = await getServerContent()` so that below you can just use the variables `comments` and `totalComments` to make it more readable.

**Review by cftorres — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/data/CommentRequest.java (hunk 19 above)
Comment (keelywan) [reply portfolio/src/main/java/com/google/sps/data/CommentRequest.java]: Yep, you're right! I'll update the description then :)

File: portfolio/src/main/java/com/google/sps/data/CommentRequest.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+package com.google.sps.data;
+
+import com.google.sps.data.Comment;
+import java.util.List;
+
+
+/** Class containing requested comment data. */ 
+public class CommentRequest {
+
+  /** List containing a strict subset of the comments in Datastore that satisfy the query
```
Comment (nickKelman) [inline portfolio/src/main/java/com/google/sps/data/CommentRequest.java]: super nit: prefer to see multi-line comments like this start on the next line

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 23]
```diff
@@ -61,9 +82,9 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get the input from the form.
     String comment = request.getParameter("user-comment");
-    String name = request.getParameter("name").trim();
+    String name = request.getParameter("name").trim().toUpperCase();
     if(name.equals("")) {
-      name = "Anonymous";
+      name = "ANONYMOUS";
```
Comment (nickKelman) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: this is another key word that could be extracted into an enum

**Review by nickKelman — APPROVED** [review]
Nice work. Remember to update the screenshots with any updated changes. I think the words that appear in the dropdown options no longer appear?
